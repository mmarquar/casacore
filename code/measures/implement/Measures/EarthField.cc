//# EarthField.cc:  EarthField class
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#ifdef __GNUG__
#include <aips/Measures/Quantum.h>
typedef Quantum<Double> gpp_mvdoppler_bug1;
#endif
#include <trial/Measures/EarthField.h>
#include <aips/Measures/MeasTable.h>
#include <aips/Measures/MVPosition.h>
#include <aips/Arrays/Vector.h>
#include <aips/Mathematics/Constants.h>
#include <aips/Mathematics/Math.h>
#include <aips/Tasking/AipsrcValue.h>

//# Constants
const Double EarthField::INTV = 50000;

//# Static data
uInt EarthField::interval_reg = 0;

//# Constructors
EarthField::EarthField() :
  method_p(EarthField::STANDARD), fixedEpoch_p(MeasData::MJD2000), agh_p(0), 
  p_p(0), q_p(0), cl_p(0), sl_p(0),
  lres(0) {
    fillField();
}

EarthField::EarthField(const EarthField &other) {
  copy(other);
}

EarthField::EarthField(EarthFieldTypes type, Double catepoch) :
  method_p(type), fixedEpoch_p(catepoch),
  p_p(0), q_p(0), cl_p(0), sl_p(0),
  lres(0) {
    fillField();
  }

EarthField &EarthField::operator=(const EarthField &other) {
  if ( this != &other) copy(other);
  return *this;
}

void EarthField::init() {
  method_p = EarthField::STANDARD;
  fixedEpoch_p = MeasData::MJD2000;
  fillField();
}

void EarthField::init(EarthFieldTypes type, Double catepoch) {
  method_p = type;
  fixedEpoch_p = catepoch;
  fillField();
}

//# Destructor
EarthField::~EarthField() {}

//# Operators
// Calculate EarthField components
const Vector<Double> &EarthField::operator()(const MVPosition &pos) {
  calcField(pos);
  Vector<Double> dx((pos-checkPos).getValue());
  lres++; lres %= 4;
  for (Int i=0; i<3; i++) {
    result[lres](i) = pval[i] +
      dx(0)*dval[0][i] + dx(1)*dval[1][i] + dx(2)*dval[2][i];
  };
  return result[lres];
}

//# Member functions
const Vector<Double> *EarthField::derivative(const MVPosition &pos) {
  calcField(pos);
  lres=0;		// Make sure contiguous set
  for (Int j=0; j<3; j++) {
    lres++; lres %= 4;
    for (Int i=0; i<3; i++) {
      result[lres](i) = dval[j][i];
    };
  };
  return &result[1];
}

void EarthField::copy(const EarthField &other) {
  method_p = other.method_p;
  fixedEpoch_p = other.fixedEpoch_p;
  agh_p = other.agh_p;
  checkPos = other.checkPos;
  for (Int i=0; i<3; i++) {
    pval[i] = other.pval[i];
    for (Int k=0; k<3; k++) dval[i][k] = other.dval[i][k];
  };
  for (Int j=0; j<4; j++) {
    result[j] = other.result[j];
  };
}

void EarthField::fillField() {

  // Get the interpolation interval
  if (!EarthField::interval_reg) {
    interval_reg = 
      AipsrcValue<Double>::registerRC(String("measures.earthfield.d_interval"),
				      Unit("km"), Unit("m"),
				      EarthField::INTV);
  };

  checkPos = MVPosition(1e30, 1e30, 1e30);
  switch (method_p) {
  default:
    agh_p.resize(0);
    agh_p =  MeasTable::IGRF(fixedEpoch_p);
    p_p.resize(PQ_LEN);
    q_p.resize(PQ_LEN);
    cl_p.resize(2*PQ_LEN);
    sl_p.resize(2*PQ_LEN);
    break;
  };
  for (Int j=0; j<4; j++) {
    result[j].resize(3);
  };

}

void EarthField::refresh() {
  fillField();
}

void EarthField::calcField(const MVPosition &pos) {
  if (!pos.nearAbs(checkPos,
		   AipsrcValue<Double>::get(EarthField::interval_reg))) {
    checkPos = pos;
    Vector<Double> posmv(3);
    posmv = pos.getValue();
    Vector<Double> posv(3);
    posv = pos.get();
    switch (method_p) {
    default: {
      Double slat, clat, x, y, z, ratio, rr, one, two, three;
      Int l, m, n, fn, fm, j, i;
      for (Int lp=0; lp<4; lp++) {
	slat = cos(C::pi_2 - posv(2));
	clat = sin(C::pi_2 - posv(2));
	cl_p(0) = cos(posv(1));
	sl_p(0) = sin(posv(1));
	x = 0.0;
	y = 0.0;
	z = 0.0;
	l = 0;
	m = 0;
	n = 0;
	ratio = 6371200/posv(0);
	//
	// Compute Schmidt quasi-normal coefficients P and X (=Q)
	//
	p_p(0) = 2.0 * slat;
	p_p(1) = 2.0 * clat;
	p_p(2) = 4.5 * slat * slat - 1.5;
	p_p(3) = 5.1961524 * clat * slat;
	q_p(0) = -clat;
	q_p(1) = slat;
	q_p(2) = -3.0 * clat * slat;
	q_p(3) = 1.7320508 * (slat * slat - clat * clat);
	
	for (Int k=0; k<PQ_LEN; k++) {
	  if (n-m-1 < 0) {
	    m = -1;
	    n++;
	    rr = pow(ratio, Double(n+2));
	    fn = n;
	  };
	  fm = m+1;
	  if (k-4 >=0) {
	    if (m+1-n == 0) {
	      one = sqrt(1.0 - 0.5/fm);
	      j = k - n - 1;
	      p_p(k) = (1.0 + 1.0/fm) * one * clat * p_p(j);
	      q_p(k) = one * (clat * q_p(j) + slat/fm * p_p(j));
	      sl_p(m) = sl_p(m-1) * cl_p(0) + cl_p(m-1) * sl_p(0);
	      cl_p(m) = cl_p(m-1) * cl_p(0)-sl_p(m-1) * sl_p(0);
	    } else {
	      one = sqrt(fn * fn - fm * fm);
	      two = sqrt((fn-1.0) * (fn-1.0) - fm * fm)/one;
	      three = (2.0 * fn - 1.0)/one;
	      i = k-n;
	      j = k - 2 * n + 1;
	      p_p(k) = (fn+1.0) * (three * slat/fn * p_p(i) -
				   two/(fn-1.0) * p_p(j));
	      q_p(k) = three * (slat * q_p(i) - clat/fn * p_p(i)) -
		two * q_p(j);
	    };
	  };
	  //
	  // Synthesise X,Y,Z in geocentric coordinates
	  //
	  one = (agh_p(l)) * rr;
	  if (m == -1) {
	    x = x + one * q_p(k);
	    z = z - one * p_p(k);
	    l++;
	  } else {
	    two = (agh_p(l+1)) * rr;
	    three = one * cl_p(m) + two * sl_p(m);
	    x = x + three * q_p(k);
	    z = z - three * p_p(k);
	    if (clat > 0) {
	      y = y + (one * sl_p(m) -
		       two * cl_p(m)) * fm * p_p(k)/((fn+1.0) * clat);
	    } else {
	      y = y + (one * sl_p(m) -
		       two * cl_p(m)) * q_p(k) * slat;
	    };
	    l += 2;
	  };
	  m++;
	}; // calculation loop
	if (lp == 0) {
	  pval[0] = x;
	  pval[1] = y;
	  pval[2] = z;
	} else {
	  dval[lp-1][0] = (x-pval[0])/DER_INTV;
	  dval[lp-1][1] = (y-pval[1])/DER_INTV;
	  dval[lp-1][2] = (z-pval[2])/DER_INTV;
	};
	if (lp < 3) {
	  if (lp != 0) posmv(lp-1) -= DER_INTV;
	  posmv(lp)   += DER_INTV;
	  posv = MVPosition(posmv).get();
	};
      }; // derivative loop
    }
    break;
    };
  };
}
