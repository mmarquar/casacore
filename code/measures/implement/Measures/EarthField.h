//# EarthField.h: EarthField class
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$

#if !defined(AIPS_EARTHFIELD_H)
#define AIPS_EARTHFIELD_H

//# Includes
#include <aips/aips.h>
#include <aips/Arrays/Vector.h>
#include <aips/Measures/MVPosition.h>

//# Forward Declarations

//# Constants
// Length of P and Q arrays, half length of CL/SL arrays
const Int PQ_LEN = 65;
// Interval (m) for derivatives
const Double DER_INTV = 10000;

// <summary> EarthField class and calculations </summary>

// <use visibility=local>

// <reviewed reviewer="" date="yyyy/mm/dd" tests="tEarthField" demos="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class=Measure>Measure</linkto> class for use
//   <li> <linkto class=MeasTable>MeasTable</linkto> class for data
// </prerequisite>
//
// <etymology>
// Earth magnetic Field
// </etymology>
//
// <synopsis>
// EarthField forms the class for Earth magnetic field calculations. It is a 
// simple container with the selected method, and the mean epoch.<br>
// The method is selected from one of the following:
// <ul>
//   <li> EarthField::STANDARD  (at 1998/05/18 the IGRF definition)
//   <li> EarthField::IGRF	(IGRF reference field)
// </ul>
// Epochs can be specified as the MJD (with defined constants
// MeasData::MJD2000 and MeasData::MJD1950 or the actual MJD),
// leading to the following constructors:
// <ul>
//   <li> EarthField() default; assuming IGRF and MJD2000
//   <li> EarthField(method); assuming J2000 as epoch
//   <li> EarthField(method, epoch) with epoch Double(MJD)
// </ul>
// Actual EarthField for a certain position on Earth is calculated by the () 
// operator. Arguments can be:
// <ul>
//   <li> MVPosition: a position on Earth (ITRF)
// </ul>
// The returned value is a 3D vector of the field (nT) in ITRF coordinates.
// The derivative (d<sup>-1</sup>) can be obtained as well by 
// derivative(MVPosition). <br>
// An EarthField can be re-initialed with a different method and/or other
// epoch with the <src>init()</src> functions (same format as constructors).
//
// To bypass the full, lengthy calculation actual returned values are calculated
// using the derivative if within about 50 km (error less than about
// 10<sup>-2</sup> G). A call to refresh() will re-initiate calculations
// from scratch.<br>
// The following details can be set with the 
// <linkto class=Aipsrc>Aipsrc</linkto> mechanism:
// <ul>
//  <li> measures.earthfield.d_interval: approximation radius
//	(km is default unit) over which linear approximation
//	is used
// </ul>
// The field is assumed to be constant over the time-span the class is used.
//
// The calculations are based on a routine provided by the IGRF community. See
// ftp.ngdc.noaa.gov/Solid_Earth/Mainfld_Mag/Models/IAGA, routine IGRFLIB.FOR.
// The values are in nT (10uG).
// </synopsis>
//
// <example>
//  <srcblock>
//	EarthField mine(EarthField::STANDARD,
//                      45837.0);		// define EarthField type
//						// for 84/05/17
//	MPosition pos;
//      MeasTable::Observatory(pos, "WSRT");	// Obervatory position
//  // Make sure correct data
//      MVPosition x(MPosition::Convert(pos, MPosition::ITRF)().getValue());
//	MVEarthMagnetic now = mine(x);		// get EarthField
//  </srcblock>
// </example>
//
// <motivation>
// To have a history container for field calculations
// </motivation>
//
// <todo asof="1998/05/18">
//   <li> nothing I know off
// </todo>

class EarthField {

public:

  //# Constants
  // Default interval to be used for linear approximation (in m)
  static const Double INTV;

  //# Enumerations
  // Types of known EarthField calculations
  enum EarthFieldTypes {
    IGRF,
    NONE,
    STANDARD = IGRF
  };

  //# Constructors
  // Default constructor, generates default J2000 EarthField identification
  EarthField();
  // Copy constructor
  EarthField(const EarthField &other);
  // Constructor with epoch in MJulian days
  explicit EarthField(EarthFieldTypes type, Double catepoch=51544.5);
  // Copy assignment
  EarthField &operator=(const EarthField &other);
  
  //# Destructor
  ~EarthField();
  
  //# Operators
  // Return the EarthField angles
  const Vector<Double> &operator()(const MVPosition &pos);
  
  //# General Member Functions
  // Return derivatives of field (to X, Y, Z)
  const Vector<Double> *derivative(const MVPosition &pos);
  // Re-initialise EarthField object
  // <group>
  void init();
  void init(EarthFieldTypes type, Double catepoch=51544.5);
  // </group>
  // Refresh calculations
  void refresh();
  
  private:
  //# Data members
  // Method to be used
  EarthFieldTypes method_p;
  // Fixed epoch to be used (MJD)
  Double fixedEpoch_p;
  // List of spherical components
  Vector<Double> agh_p;
  // Work arrays for calculations
  // <group>
  Vector<Double> p_p;
  Vector<Double> q_p;
  Vector<Double> cl_p;
  Vector<Double> sl_p;
  // </group>
  // Check position
  MVPosition checkPos;
  // Cached calculated field components
  Double pval[3];
  // Cached derivatives
  Double dval[3][3];
  // To reference results, and use a few in interim calculations, results are
  // calculated in a circular buffer.
  // Current result pointer
  Int lres;
  // Last calculation
  Vector<Double> result[4];
  // Interpolation interval
  static uInt interval_reg;

  //# Member functions
  // Make a copy
  void copy(const EarthField &other);
  // Create correct default fixedEpoch and catalogue field data
  void fillField();
  // Calculate EarthField for longitude and latitude and altitude (m)
  void calcField(const MVPosition &pos);

};
  
#endif


