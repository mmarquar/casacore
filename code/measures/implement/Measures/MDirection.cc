//# MDirection.cc:  A Measure: astronomical direction
//# Copyright (C) 1995,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#ifdef __GNUG__
#include <aips/Measures/Quantum.h>
typedef Quantum<Double> gpp_direction_bug1;
#endif
#include <aips/Exceptions.h>
#include <aips/Utilities/Assert.h>
#include <aips/Arrays/Vector.h>
#include <aips/RTTI/Register.h>
#include <aips/Measures/MDirection.h>

//# Constructors
MDirection::MDirection() :
  MeasBase<MVDirection,MDirection::Ref>() {}

MDirection::MDirection(const MVDirection &dt) : 
  MeasBase<MVDirection,MDirection::Ref>(dt,MDirection::DEFAULT) {}

MDirection::MDirection(const MVDirection &dt, const MDirection::Ref &rf) : 
  MeasBase<MVDirection,MDirection::Ref>(dt,rf) {}

MDirection::MDirection(const MVDirection &dt, uInt rf) : 
  MeasBase<MVDirection,MDirection::Ref>(dt,rf) {}

MDirection::MDirection(const Quantity &dt, const Quantity &dt1) : 
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt,dt1),
					MDirection::DEFAULT) {}

MDirection::MDirection(const Quantity &dt, const Quantity &dt1,
		       const MDirection::Ref &rf) : 
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt,dt1),rf) {}

MDirection::MDirection(const Quantity &dt, const Quantity &dt1,
		       uInt rf) : 
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt,dt1),rf) {}

MDirection::MDirection(const Quantum<Vector<Double> > &dt) :
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt),
					MDirection::DEFAULT) {}

MDirection::MDirection(const Quantum<Vector<Double> > &dt,
		       const MDirection::Ref &rf) : 
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt),rf) {}

MDirection::MDirection(const Quantum<Vector<Double> > &dt,
		       uInt rf) : 
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt),rf) {}

MDirection::MDirection(const Measure *dt) :
  MeasBase<MVDirection,MDirection::Ref>(dt) {}

MDirection::MDirection(const MeasValue *dt) :
  MeasBase<MVDirection,MDirection::Ref>(*(MVDirection*)dt,
					MDirection::DEFAULT) {}

MDirection::MDirection(const MDirection::Ref &rf) : 
  MeasBase<MVDirection,MDirection::Ref>(rf) {}

//# Destructor
MDirection::~MDirection() {}

//# Operators

//# Member functions

const String &MDirection::tellMe() const {
    return MDirection::showMe();
}

const String &MDirection::showMe() {
    static const String name("Direction");
    return name;
}

uInt MDirection::type() const {
  return Register((MDirection *)0);
}

void MDirection::assert(const Measure &in) {
  if (in.type() != Register((MDirection *)0)) {
    throw(AipsError("Illegal Measure type argument: " +
		    MDirection::showMe()));
  };
}

const String &MDirection::showType(uInt tp) {
    static const String tname[MDirection::N_Types] = {
	"J2000",
	"JMEAN",
	"JTRUE",
	"APP",
	"B1950",
	"BMEAN",
	"BTRUE",
	"GALACTIC",
	"HADEC",
	"AZEL",
        "AZELSW",
	"JNAT",
	"ECLIPTIC",
	"MECLIPTIC",
	"TECLIPTIC",
	"SUPERGAL" };
    static const String pname[MDirection::N_Planets - MDirection::MERCURY] = {
	"MERCURY",
	"VENUS",
	"MARS",
	"JUPITER",
	"SATURN",
	"URANUS",
	"NEPTUNE",
	"PLUTO",
	"SUN",
	"MOON" };

    if ((tp & MDirection::EXTRA) == 0) {
      DebugAssert(tp < MDirection::N_Types, AipsError);
      return tname[tp];
    };
    DebugAssert((tp & ~MDirection::EXTRA) < 
		(MDirection::N_Planets - MDirection::MERCURY), AipsError);
    return pname[tp & ~MDirection::EXTRA];
}

Bool MDirection::giveMe(const String &in, MDirection::Ref &mr) {
    static const Int N_name = 27;
    static const String tname[N_name] = {
	"J2000",
	"JMEAN",
	"JTRUE",
	"APP",
	"B1950",
	"BMEAN",
	"BTRUE",
	"GALACTIC",
	"HADEC",
	"AZEL",
        "AZELSW",
	"AZELNE",
	"JNAT",
	"ECLIPTIC",
	"MECLIPTIC",
	"TECLIPTIC",
	"SUPERGAL",
	"MERCURY",
	"VENUS",
	"MARS",
	"JUPITER",
	"SATURN",
	"URANUS",
	"NEPTUNE",
	"PLUTO",
	"SUN",
	"MOON" };

    static const uInt oname[N_name] = {
	MDirection::J2000,
	MDirection::JMEAN,
	MDirection::JTRUE,
	MDirection::APP,
	MDirection::B1950,
	MDirection::BMEAN,
	MDirection::BTRUE,
	MDirection::GALACTIC,
	MDirection::HADEC,
	MDirection::AZEL,
        MDirection::AZELSW,
	MDirection::AZEL,
	MDirection::JNAT,
	MDirection::ECLIPTIC,
	MDirection::MECLIPTIC,
	MDirection::TECLIPTIC,
	MDirection::SUPERGAL,
	MDirection::MERCURY,
	MDirection::VENUS,
	MDirection::MARS,
	MDirection::JUPITER,
	MDirection::SATURN,
	MDirection::URANUS,
	MDirection::NEPTUNE,
	MDirection::PLUTO,
	MDirection::SUN,
	MDirection::MOON };

    uInt i = Measure::giveMe(in, N_name, tname);

    if (i>=N_name) {
	mr = MDirection::Ref();
	return False;
    } else {
	mr = MDirection::Ref(oname[i]);
    };
    return True;
}

MDirection::GlobalTypes MDirection::globalType(uInt tp) {

    static const MDirection::GlobalTypes oname[MDirection::N_Types] = {
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GLONGLAT,
	MDirection::GHADEC,
	MDirection::GAZEL,
        MDirection::GAZEL,
	MDirection::GRADEC,
	MDirection::GLONGLAT,
	MDirection::GLONGLAT,
	MDirection::GLONGLAT,
	MDirection::GLONGLAT };
    if ((tp & MDirection::EXTRA) != 0) tp = 0;
    DebugAssert(tp < MDirection::N_Types, AipsError);

    return oname[tp];
}

Quantum<Vector<Double> > MDirection::getAngle() const {
    return (data.getAngle());
}

Quantum<Vector<Double> > MDirection::getAngle(const Unit &inunit) const {
    return (data.getAngle(inunit));
}

Measure *MDirection::clone() const {
    return (new MDirection(*this));
}
