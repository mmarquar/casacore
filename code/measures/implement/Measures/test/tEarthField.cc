//# tMEarthMagnetic.cc: This program test Measure functions
//# Copyright (C) 1998,1999,2000,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <aips/aips.h>
#include <aips/Exceptions/Error.h>
#include <aips/Measures.h>
#include <aips/Measures/EarthField.h>
#include <aips/Quanta/MVTime.h>
#include <aips/Measures/MDirection.h>
#include <aips/Measures/MPosition.h>
#include <aips/Measures/MEpoch.h>
#include <aips/Measures/MeasTable.h>
#include <aips/Arrays/Vector.h>
#include <aips/iostream.h>

int main() {

    try {
	cout << "Test Earth Magnetic field values" << endl;
	cout << "--------------------------------------" << endl;
	MVTime dat(1998,5,18);
	MVPosition mvobs(Quantity(3828488.86, "m").getBaseValue(),
			 Quantity(443253.42, "m").getBaseValue(),
			 Quantity(5064977.78, "m").getBaseValue());
	MPosition obs(mvobs);
	MeasFrame frame((MEpoch(MVEpoch(dat.day()))), obs);

	cout << "Date:      " << dat.string(MVTime::YMD +
					    MVTime::NO_TIME, 6) <<
	  endl;
	cout << "Position:  " << obs.getValue().get() << endl;
	cout << "           " << obs.getAngle("deg") << endl;

	cout << "----- IGRF coefficients" << endl;
	cout << "Field      " <<
	  MeasTable::IGRF(dat.day()) << endl;
	cout << "             " <<
	  MeasTable::IGRF(dat.day()).nelements() << endl;

	EarthField ef(EarthField::STANDARD, dat.day());
	cout << "Result:    " << ef(obs.getValue()) << endl;
	cout << "Derivatives: " << endl;
	for (Int i0=0; i0<3; i0++) {
	  cout << "     " << ef.derivative(obs.getValue())[i0] << endl;
	};

	cout << "--------- From derivatives ----------" << endl;
	cout << "+10km X:   " <<
	  ef((obs.getValue()+MVPosition(10000,0,0))) << endl;
	cout << "-10km X:   " <<
	  ef((obs.getValue()+MVPosition(-10000,0,0))) << endl;
	cout << "+10km Y:   " <<
	  ef((obs.getValue()+MVPosition(0,10000,0))) << endl;
	cout << "-10km Y:   " <<
	  ef((obs.getValue()+MVPosition(0,-10000,0))) << endl;
	cout << "+10km Z:   " <<
	  ef((obs.getValue()+MVPosition(0,0,10000))) << endl;
	cout << "-10km Z:   " <<
	  ef((obs.getValue()+MVPosition(0,0,-10000))) << endl;

	cout << "--------- From scratch ----------" << endl;
	ef.init(EarthField::STANDARD, dat.day());
	cout << "+10km X:   " <<
	  ef((obs.getValue()+MVPosition(10000,0,0))) << endl;
	ef.init(EarthField::STANDARD, dat.day());
	cout << "-10km X:   " <<
	  ef((obs.getValue()+MVPosition(-10000,0,0))) << endl;
	ef.init(EarthField::STANDARD, dat.day());
	cout << "+10km Y:   " <<
	  ef((obs.getValue()+MVPosition(0,10000,0))) << endl;
	ef.init(EarthField::STANDARD, dat.day());
	cout << "-10km Y:   " <<
	  ef((obs.getValue()+MVPosition(0,-10000,0))) << endl;
	ef.init(EarthField::STANDARD, dat.day());
	cout << "+10km Z:   " <<
	  ef((obs.getValue()+MVPosition(0,0,10000))) << endl;
	ef.init(EarthField::STANDARD, dat.day());
	cout << "-10km Z:   " <<
	  ef((obs.getValue()+MVPosition(0,0,-10000))) << endl;

	cout << "--------- From refresh ----------" << endl;
	ef.refresh();
	cout << "+10km X:   " <<
	  ef((obs.getValue()+MVPosition(10000,0,0))) << endl;
	ef.refresh();
	cout << "-10km X:   " <<
	  ef((obs.getValue()+MVPosition(-10000,0,0))) << endl;
	ef.refresh();
	cout << "+10km Y:   " <<
	  ef((obs.getValue()+MVPosition(0,10000,0))) << endl;
	ef.refresh();
	cout << "-10km Y:   " <<
	  ef((obs.getValue()+MVPosition(0,-10000,0))) << endl;
	ef.refresh();
	cout << "+10km Z:   " <<
	  ef((obs.getValue()+MVPosition(0,0,10000))) << endl;
	ef.refresh();
	cout << "-10km Z:   " <<
	  ef((obs.getValue()+MVPosition(0,0,-10000))) << endl;

	cout << "------------------------------------------" << endl;

    } catch (AipsError x) {
	cout << x.getMesg() << endl;
    } 

    exit(0);
}
