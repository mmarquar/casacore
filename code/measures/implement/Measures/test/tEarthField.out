Test Earth Magnetic field values
--------------------------------------
Date:      1998/05/18
Position:  [6.36457e+06, 0.115264, 0.92034]
           [6.60417, 52.7316] deg
----- IGRF coefficients
Field      [-29622.6, -1745.14, 5256.26, -2241.53, 3086.48, -2406.61, 1682.3, -454.689, 1334.06, -2289.59, -249.168, 1248.33, 309.422, 741.673, -446.822, 943.699, 785.036, 268.073, 267.721, -227.952, -419.313, 107.109, 100.481, -304.374, -207.301, 352.337, 44.6747, 231.939, 161.048, -128.747, -150.988, -167.337, -57.9273, -18.2405, 102.036, 67.6869, 62.6505, -14.9879, 67.0242, 71.6021, -165.59, 66.3253, 1.32526, -60.0363, 16.3253, 7.37372, -94, 35.4222, 77.3253, -69.699, -74.301, -1.02423, -24.3253, 31.0242, 5.02423, 8.04846, 20.6505, 8.33737, 16, 10.6747, -24.0121, -4.02423, -3, 25.0121, 3.32526, 13.3495, -0.662628, -20.6747, -7.65051, 7.67474, -17.7111, -18.6384, 5.01211, 12, 5.67474, 5.95154, -3.03634, -19.3616, -8.01211, -12.0242, 4, 9, -19, 1, 15, -12, 11, 9, -7, -4, -7, -2, 9, 7, 7, 0, -8, -6, 1, -3, -4, 2, 2, 1, -5, 3, -2, 6, 4, -4, 3, 0, 1, -2, 3, 3, 3, -1, 0, -6]
             120
Result:    [12747.6, 1091.33, 47047]
Derivatives: 
     [-0.00314578, -0.00108743, -0.0146417]
     [-0.000363059, 0.005059, -0.00133054]
     [-0.00606614, -0.00128894, -0.01492]
--------- From derivatives ----------
+10km X:   [12716.1, 1080.46, 46900.6]
-10km X:   [12779, 1102.2, 47193.4]
+10km Y:   [12744, 1141.92, 47033.7]
-10km Y:   [12751.2, 1040.74, 47060.3]
+10km Z:   [12686.9, 1078.44, 46897.8]
-10km Z:   [12808.2, 1104.22, 47196.2]
--------- From scratch ----------
+10km X:   [12716.1, 1080.46, 46900.6]
-10km X:   [12778.9, 1102.24, 47193.7]
+10km Y:   [12744, 1141.92, 47033.7]
-10km Y:   [12751, 1040.63, 47060]
+10km Z:   [12686.9, 1078.44, 46897.8]
-10km Z:   [12808.6, 1104.34, 47196.8]
--------- From refresh ----------
+10km X:   [12716.1, 1080.46, 46900.6]
-10km X:   [12778.9, 1102.24, 47193.7]
+10km Y:   [12744, 1141.92, 47033.7]
-10km Y:   [12751, 1040.63, 47060]
+10km Z:   [12686.9, 1078.44, 46897.8]
-10km Z:   [12808.6, 1104.34, 47196.8]
------------------------------------------
